Trouble implementing AsyncIterator interface in 3.6.2

<!-- 🚨 STOP 🚨 𝗦𝗧𝗢𝗣 🚨 𝑺𝑻𝑶𝑷 🚨

Half of all issues filed here are duplicates, answered in the FAQ, or not appropriate for the bug tracker. Even if you think you've found a *bug*, please read the FAQ first, especially the Common "Bugs" That Aren't Bugs section!

Please help us by doing the following steps before logging an issue:
  * Search: https://github.com/Microsoft/TypeScript/search?type=Issues
  * Read the FAQ: https://github.com/Microsoft/TypeScript/wiki/FAQ

Please fill in the *entire* template below.
-->

<!-- Please try to reproduce the issue with `typescript@next`. It may have already been fixed. -->
**TypeScript Version:**  3.6.2

<!-- Search terms you tried before logging this (so others can find this issue more easily) -->
**Search Terms:**
async iterator iterable next

**Code:**
filename: constant.ts
```ts
class ConstantIterator<T> implements AsyncIterator<T, undefined, T | undefined> {
  constructor(private constant: T) {
  }
  async next(value?: T): Promise<IteratorResult<T>> {
    if (value != null) {
      throw new Error("ConstantIterator.prototype.next may not take any values");
    }
    return {value: this.constant, done: false};
  }
}
```

**Expected behavior:**
Code compiles.
**Actual behavior:**
```js
constant.ts(4,9): error TS2416: Property 'next' in type 'ConstantIterator<T>' is not assignable to the same property in base type 'AsyncIterator<T, undefined, T>'.
  Type '(value?: T) => Promise<{ value: T; done: boolean; }>' is not assignable to type '(...args: [] | [T | PromiseLike<T>]) => Promise<IteratorResult<T, undefined>>'.
    Types of parameters 'value' and 'args' are incompatible.
      Type '[] | [T | PromiseLike<T>]' is not assignable to type '[T?]'.
        Type '[T | PromiseLike<T>]' is not assignable to type '[T?]'.
          Types of property '0' are incompatible.
            Type 'T | PromiseLike<T>' is not assignable to type 'T'.
              'T | PromiseLike<T>' is assignable to the constraint of type 'T', but 'T' could be instantiated with a different subtype of constraint '{}'.
                Type 'PromiseLike<T>' is not assignable to type 'T'.
                  'PromiseLike<T>' is assignable to the constraint of type 'T', but 'T' could be instantiated with a different subtype of constraint '{}'.
```

Note that the synchronous version of `ConstantIterator` does not throw the same error. It’s the type union of `T | PromiseLike<T>` which seems to cause the error.

**Playground Link:** <!-- A link to a TypeScript Playground "Share" link which demonstrates this behavior -->
Playground has yet to be updated for 3.6.

**Related Issues:** <!-- Did you find other bugs that looked similar? -->
https://github.com/microsoft/TypeScript/pull/30790 Strongly typed iterator PR.
https://github.com/microsoft/TypeScript/issues/32682 Other (async) iterator usability issues.
https://github.com/microsoft/TypeScript/issues/33131 Possible change to type inference in 3.6.

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -28740,11 +28740,13 @@ namespace ts {
             let nextType: Type | undefined;
             if (methodName !== "throw") {
                 const methodParameterType = methodParameterTypes ? getUnionType(methodParameterTypes) : unknownType;
-                const resolvedMethodParameterType = resolver.resolveIterationType(methodParameterType, errorNode) || anyType;
                 if (methodName === "next") {
-                    nextType = resolvedMethodParameterType;
+                    // The value of `next(value)` is *not* awaited by async generators
+                    nextType = methodParameterType;
                 }
                 else if (methodName === "return") {
+                    // The value of `return(value)` *is* awaited by async generators
+                    const resolvedMethodParameterType = resolver.resolveIterationType(methodParameterType, errorNode) || anyType;
                     returnTypes = append(returnTypes, resolvedMethodParameterType);
                 }
             }

--- a/src/lib/es2018.asyncgenerator.d.ts
+++ b/src/lib/es2018.asyncgenerator.d.ts
@@ -2,7 +2,7 @@
 
 interface AsyncGenerator<T = unknown, TReturn = any, TNext = unknown> extends AsyncIterator<T, TReturn, TNext> {
     // NOTE: 'next' is defined using a tuple to ensure we report the correct assignability errors in all places.
-    next(...args: [] | [TNext | PromiseLike<TNext>]): Promise<IteratorResult<T, TReturn>>;
+    next(...args: [] | [TNext]): Promise<IteratorResult<T, TReturn>>;
     return(value: TReturn | PromiseLike<TReturn>): Promise<IteratorResult<T, TReturn>>;
     throw(e: any): Promise<IteratorResult<T, TReturn>>;
     [Symbol.asyncIterator](): AsyncGenerator<T, TReturn, TNext>;

--- a/src/lib/es2018.asynciterable.d.ts
+++ b/src/lib/es2018.asynciterable.d.ts
@@ -11,7 +11,7 @@ interface SymbolConstructor {
 
 interface AsyncIterator<T, TReturn = any, TNext = undefined> {
     // NOTE: 'next' is defined using a tuple to ensure we report the correct assignability errors in all places.
-    next(...args: [] | [TNext | PromiseLike<TNext>]): Promise<IteratorResult<T, TReturn>>;
+    next(...args: [] | [TNext]): Promise<IteratorResult<T, TReturn>>;
     return?(value?: TReturn | PromiseLike<TReturn>): Promise<IteratorResult<T, TReturn>>;
     throw?(e?: any): Promise<IteratorResult<T, TReturn>>;
 }

--- a/tests/baselines/reference/customAsyncIterator.js
+++ b/tests/baselines/reference/customAsyncIterator.js
@@ -1,0 +1,26 @@
+//// [customAsyncIterator.ts]
+// GH: https://github.com/microsoft/TypeScript/issues/33239
+class ConstantIterator<T> implements AsyncIterator<T, undefined, T | undefined> {
+    constructor(private constant: T) {
+    }
+    async next(value?: T): Promise<IteratorResult<T>> {
+        if (value != null) {
+            throw new Error("ConstantIterator.prototype.next may not take any values");
+        }
+        return { value: this.constant, done: false };
+    }
+}
+
+//// [customAsyncIterator.js]
+// GH: https://github.com/microsoft/TypeScript/issues/33239
+class ConstantIterator {
+    constructor(constant) {
+        this.constant = constant;
+    }
+    async next(value) {
+        if (value != null) {
+            throw new Error("ConstantIterator.prototype.next may not take any values");
+        }
+        return { value: this.constant, done: false };
+    }
+}

--- a/tests/baselines/reference/customAsyncIterator.symbols
+++ b/tests/baselines/reference/customAsyncIterator.symbols
@@ -1,0 +1,35 @@
+=== tests/cases/compiler/customAsyncIterator.ts ===
+// GH: https://github.com/microsoft/TypeScript/issues/33239
+class ConstantIterator<T> implements AsyncIterator<T, undefined, T | undefined> {
+>ConstantIterator : Symbol(ConstantIterator, Decl(customAsyncIterator.ts, 0, 0))
+>T : Symbol(T, Decl(customAsyncIterator.ts, 1, 23))
+>AsyncIterator : Symbol(AsyncIterator, Decl(lib.es2018.asynciterable.d.ts, --, --))
+>T : Symbol(T, Decl(customAsyncIterator.ts, 1, 23))
+>T : Symbol(T, Decl(customAsyncIterator.ts, 1, 23))
+
+    constructor(private constant: T) {
+>constant : Symbol(ConstantIterator.constant, Decl(customAsyncIterator.ts, 2, 16))
+>T : Symbol(T, Decl(customAsyncIterator.ts, 1, 23))
+    }
+    async next(value?: T): Promise<IteratorResult<T>> {
+>next : Symbol(ConstantIterator.next, Decl(customAsyncIterator.ts, 3, 5))
+>value : Symbol(value, Decl(customAsyncIterator.ts, 4, 15))
+>T : Symbol(T, Decl(customAsyncIterator.ts, 1, 23))
+>Promise : Symbol(Promise, Decl(lib.es5.d.ts, --, --), Decl(lib.es2015.iterable.d.ts, --, --), Decl(lib.es2015.promise.d.ts, --, --), Decl(lib.es2015.symbol.wellknown.d.ts, --, --), Decl(lib.es2018.promise.d.ts, --, --))
+>IteratorResult : Symbol(IteratorResult, Decl(lib.es2015.iterable.d.ts, --, --))
+>T : Symbol(T, Decl(customAsyncIterator.ts, 1, 23))
+
+        if (value != null) {
+>value : Symbol(value, Decl(customAsyncIterator.ts, 4, 15))
+
+            throw new Error("ConstantIterator.prototype.next may not take any values");
+>Error : Symbol(Error, Decl(lib.es5.d.ts, --, --), Decl(lib.es5.d.ts, --, --))
+        }
+        return { value: this.constant, done: false };
+>value : Symbol(value, Decl(customAsyncIterator.ts, 8, 16))
+>this.constant : Symbol(ConstantIterator.constant, Decl(customAsyncIterator.ts, 2, 16))
+>this : Symbol(ConstantIterator, Decl(customAsyncIterator.ts, 0, 0))
+>constant : Symbol(ConstantIterator.constant, Decl(customAsyncIterator.ts, 2, 16))
+>done : Symbol(done, Decl(customAsyncIterator.ts, 8, 38))
+    }
+}

--- a/tests/baselines/reference/customAsyncIterator.types
+++ b/tests/baselines/reference/customAsyncIterator.types
@@ -1,0 +1,32 @@
+=== tests/cases/compiler/customAsyncIterator.ts ===
+// GH: https://github.com/microsoft/TypeScript/issues/33239
+class ConstantIterator<T> implements AsyncIterator<T, undefined, T | undefined> {
+>ConstantIterator : ConstantIterator<T>
+
+    constructor(private constant: T) {
+>constant : T
+    }
+    async next(value?: T): Promise<IteratorResult<T>> {
+>next : (value?: T) => Promise<IteratorResult<T, any>>
+>value : T
+
+        if (value != null) {
+>value != null : boolean
+>value : T
+>null : null
+
+            throw new Error("ConstantIterator.prototype.next may not take any values");
+>new Error("ConstantIterator.prototype.next may not take any values") : Error
+>Error : ErrorConstructor
+>"ConstantIterator.prototype.next may not take any values" : "ConstantIterator.prototype.next may not take any values"
+        }
+        return { value: this.constant, done: false };
+>{ value: this.constant, done: false } : { value: T; done: false; }
+>value : T
+>this.constant : T
+>this : this
+>constant : T
+>done : false
+>false : false
+    }
+}

--- a/tests/baselines/reference/types.asyncGenerators.es2018.1.symbols
+++ b/tests/baselines/reference/types.asyncGenerators.es2018.1.symbols
@@ -289,4 +289,13 @@ async function * awaitedType2() {
 >Promise : Symbol(Promise, Decl(lib.es5.d.ts, --, --), Decl(lib.es2015.iterable.d.ts, --, --), Decl(lib.es2015.promise.d.ts, --, --), Decl(lib.es2015.symbol.wellknown.d.ts, --, --), Decl(lib.es2018.promise.d.ts, --, --))
 >resolve : Symbol(PromiseConstructor.resolve, Decl(lib.es2015.promise.d.ts, --, --), Decl(lib.es2015.promise.d.ts, --, --))
 }
+async function * nextType1(): { next(...args: [] | [number | PromiseLike<number>]): any } {
+>nextType1 : Symbol(nextType1, Decl(types.asyncGenerators.es2018.1.ts, 122, 1))
+>next : Symbol(next, Decl(types.asyncGenerators.es2018.1.ts, 123, 31))
+>args : Symbol(args, Decl(types.asyncGenerators.es2018.1.ts, 123, 37))
+>PromiseLike : Symbol(PromiseLike, Decl(lib.es5.d.ts, --, --))
+
+    const x = yield; // `number | PromiseLike<number>` (should not await TNext)
+>x : Symbol(x, Decl(types.asyncGenerators.es2018.1.ts, 124, 9))
+}
 

--- a/tests/baselines/reference/types.asyncGenerators.es2018.1.types
+++ b/tests/baselines/reference/types.asyncGenerators.es2018.1.types
@@ -430,4 +430,13 @@ async function * awaitedType2() {
 >resolve : { <T>(value: T | PromiseLike<T>): Promise<T>; (): Promise<void>; }
 >1 : 1
 }
+async function * nextType1(): { next(...args: [] | [number | PromiseLike<number>]): any } {
+>nextType1 : () => { next(...args: [] | [number | PromiseLike<number>]): any; }
+>next : (...args: [] | [number | PromiseLike<number>]) => any
+>args : [] | [number | PromiseLike<number>]
+
+    const x = yield; // `number | PromiseLike<number>` (should not await TNext)
+>x : number | PromiseLike<number>
+>yield : number | PromiseLike<number>
+}
 

--- a/tests/baselines/reference/types.asyncGenerators.es2018.2.errors.txt
+++ b/tests/baselines/reference/types.asyncGenerators.es2018.2.errors.txt
@@ -3,7 +3,7 @@ tests/cases/conformance/types/asyncGenerators/types.asyncGenerators.es2018.2.ts(
 tests/cases/conformance/types/asyncGenerators/types.asyncGenerators.es2018.2.ts(10,7): error TS2322: Type '() => AsyncGenerator<string, void, undefined>' is not assignable to type '() => AsyncIterableIterator<number>'.
   Type 'AsyncGenerator<string, void, undefined>' is not assignable to type 'AsyncIterableIterator<number>'.
     Types of property 'next' are incompatible.
-      Type '(...args: [] | [PromiseLike<undefined>]) => Promise<IteratorResult<string, void>>' is not assignable to type '(...args: [] | [PromiseLike<undefined>]) => Promise<IteratorResult<number, any>>'.
+      Type '(...args: [] | [undefined]) => Promise<IteratorResult<string, void>>' is not assignable to type '(...args: [] | [undefined]) => Promise<IteratorResult<number, any>>'.
         Type 'Promise<IteratorResult<string, void>>' is not assignable to type 'Promise<IteratorResult<number, any>>'.
           Type 'IteratorResult<string, void>' is not assignable to type 'IteratorResult<number, any>'.
             Type 'IteratorYieldResult<string>' is not assignable to type 'IteratorResult<number, any>'.
@@ -14,7 +14,7 @@ tests/cases/conformance/types/asyncGenerators/types.asyncGenerators.es2018.2.ts(
 tests/cases/conformance/types/asyncGenerators/types.asyncGenerators.es2018.2.ts(16,7): error TS2322: Type '() => AsyncGenerator<string, void, unknown>' is not assignable to type '() => AsyncIterableIterator<number>'.
   Type 'AsyncGenerator<string, void, unknown>' is not assignable to type 'AsyncIterableIterator<number>'.
     Types of property 'next' are incompatible.
-      Type '(...args: [] | [unknown]) => Promise<IteratorResult<string, void>>' is not assignable to type '(...args: [] | [PromiseLike<undefined>]) => Promise<IteratorResult<number, any>>'.
+      Type '(...args: [] | [unknown]) => Promise<IteratorResult<string, void>>' is not assignable to type '(...args: [] | [undefined]) => Promise<IteratorResult<number, any>>'.
         Type 'Promise<IteratorResult<string, void>>' is not assignable to type 'Promise<IteratorResult<number, any>>'.
 tests/cases/conformance/types/asyncGenerators/types.asyncGenerators.es2018.2.ts(19,7): error TS2322: Type '() => AsyncGenerator<string, void, undefined>' is not assignable to type '() => AsyncIterable<number>'.
   Type 'AsyncGenerator<string, void, undefined>' is not assignable to type 'AsyncIterable<number>'.
@@ -22,7 +22,7 @@ tests/cases/conformance/types/asyncGenerators/types.asyncGenerators.es2018.2.ts(
       Type '() => AsyncGenerator<string, void, undefined>' is not assignable to type '() => AsyncIterator<number, any, undefined>'.
         Type 'AsyncGenerator<string, void, undefined>' is not assignable to type 'AsyncIterator<number, any, undefined>'.
           Types of property 'next' are incompatible.
-            Type '(...args: [] | [PromiseLike<undefined>]) => Promise<IteratorResult<string, void>>' is not assignable to type '(...args: [] | [PromiseLike<undefined>]) => Promise<IteratorResult<number, any>>'.
+            Type '(...args: [] | [undefined]) => Promise<IteratorResult<string, void>>' is not assignable to type '(...args: [] | [undefined]) => Promise<IteratorResult<number, any>>'.
               Type 'Promise<IteratorResult<string, void>>' is not assignable to type 'Promise<IteratorResult<number, any>>'.
 tests/cases/conformance/types/asyncGenerators/types.asyncGenerators.es2018.2.ts(22,7): error TS2322: Type '() => AsyncGenerator<string, void, undefined>' is not assignable to type '() => AsyncIterable<number>'.
   Type 'AsyncGenerator<string, void, undefined>' is not assignable to type 'AsyncIterable<number>'.
@@ -32,7 +32,7 @@ tests/cases/conformance/types/asyncGenerators/types.asyncGenerators.es2018.2.ts(
       Type '() => AsyncGenerator<string, void, unknown>' is not assignable to type '() => AsyncIterator<number, any, undefined>'.
         Type 'AsyncGenerator<string, void, unknown>' is not assignable to type 'AsyncIterator<number, any, undefined>'.
           Types of property 'next' are incompatible.
-            Type '(...args: [] | [unknown]) => Promise<IteratorResult<string, void>>' is not assignable to type '(...args: [] | [PromiseLike<undefined>]) => Promise<IteratorResult<number, any>>'.
+            Type '(...args: [] | [unknown]) => Promise<IteratorResult<string, void>>' is not assignable to type '(...args: [] | [undefined]) => Promise<IteratorResult<number, any>>'.
               Type 'Promise<IteratorResult<string, void>>' is not assignable to type 'Promise<IteratorResult<number, any>>'.
 tests/cases/conformance/types/asyncGenerators/types.asyncGenerators.es2018.2.ts(28,7): error TS2322: Type '() => AsyncGenerator<string, void, undefined>' is not assignable to type '() => AsyncIterator<number, any, undefined>'.
   Type 'AsyncGenerator<string, void, undefined>' is not assignable to type 'AsyncIterator<number, any, undefined>'.
@@ -53,7 +53,7 @@ tests/cases/conformance/types/asyncGenerators/types.asyncGenerators.es2018.2.ts(
 tests/cases/conformance/types/asyncGenerators/types.asyncGenerators.es2018.2.ts(67,42): error TS2741: Property '[Symbol.iterator]' is missing in type 'AsyncGenerator<any, any, any>' but required in type 'Iterable<number>'.
 tests/cases/conformance/types/asyncGenerators/types.asyncGenerators.es2018.2.ts(70,42): error TS2322: Type 'AsyncGenerator<number, any, undefined>' is not assignable to type 'Iterator<number, any, undefined>'.
   Types of property 'next' are incompatible.
-    Type '(...args: [] | [PromiseLike<undefined>]) => Promise<IteratorResult<number, any>>' is not assignable to type '(...args: [] | [undefined]) => IteratorResult<number, any>'.
+    Type '(...args: [] | [undefined]) => Promise<IteratorResult<number, any>>' is not assignable to type '(...args: [] | [undefined]) => IteratorResult<number, any>'.
       Type 'Promise<IteratorResult<number, any>>' is not assignable to type 'IteratorResult<number, any>'.
         Property 'value' is missing in type 'Promise<IteratorResult<number, any>>' but required in type 'IteratorYieldResult<number>'.
 tests/cases/conformance/types/asyncGenerators/types.asyncGenerators.es2018.2.ts(74,12): error TS2504: Type '{}' must have a '[Symbol.asyncIterator]()' method that returns an async iterator.
@@ -79,7 +79,7 @@ tests/cases/conformance/types/asyncGenerators/types.asyncGenerators.es2018.2.ts(
 !!! error TS2322: Type '() => AsyncGenerator<string, void, undefined>' is not assignable to type '() => AsyncIterableIterator<number>'.
 !!! error TS2322:   Type 'AsyncGenerator<string, void, undefined>' is not assignable to type 'AsyncIterableIterator<number>'.
 !!! error TS2322:     Types of property 'next' are incompatible.
-!!! error TS2322:       Type '(...args: [] | [PromiseLike<undefined>]) => Promise<IteratorResult<string, void>>' is not assignable to type '(...args: [] | [PromiseLike<undefined>]) => Promise<IteratorResult<number, any>>'.
+!!! error TS2322:       Type '(...args: [] | [undefined]) => Promise<IteratorResult<string, void>>' is not assignable to type '(...args: [] | [undefined]) => Promise<IteratorResult<number, any>>'.
 !!! error TS2322:         Type 'Promise<IteratorResult<string, void>>' is not assignable to type 'Promise<IteratorResult<number, any>>'.
 !!! error TS2322:           Type 'IteratorResult<string, void>' is not assignable to type 'IteratorResult<number, any>'.
 !!! error TS2322:             Type 'IteratorYieldResult<string>' is not assignable to type 'IteratorResult<number, any>'.
@@ -98,7 +98,7 @@ tests/cases/conformance/types/asyncGenerators/types.asyncGenerators.es2018.2.ts(
 !!! error TS2322: Type '() => AsyncGenerator<string, void, unknown>' is not assignable to type '() => AsyncIterableIterator<number>'.
 !!! error TS2322:   Type 'AsyncGenerator<string, void, unknown>' is not assignable to type 'AsyncIterableIterator<number>'.
 !!! error TS2322:     Types of property 'next' are incompatible.
-!!! error TS2322:       Type '(...args: [] | [unknown]) => Promise<IteratorResult<string, void>>' is not assignable to type '(...args: [] | [PromiseLike<undefined>]) => Promise<IteratorResult<number, any>>'.
+!!! error TS2322:       Type '(...args: [] | [unknown]) => Promise<IteratorResult<string, void>>' is not assignable to type '(...args: [] | [undefined]) => Promise<IteratorResult<number, any>>'.
 !!! error TS2322:         Type 'Promise<IteratorResult<string, void>>' is not assignable to type 'Promise<IteratorResult<number, any>>'.
         yield* (async function * () { yield "a"; })();
     };
@@ -110,7 +110,7 @@ tests/cases/conformance/types/asyncGenerators/types.asyncGenerators.es2018.2.ts(
 !!! error TS2322:       Type '() => AsyncGenerator<string, void, undefined>' is not assignable to type '() => AsyncIterator<number, any, undefined>'.
 !!! error TS2322:         Type 'AsyncGenerator<string, void, undefined>' is not assignable to type 'AsyncIterator<number, any, undefined>'.
 !!! error TS2322:           Types of property 'next' are incompatible.
-!!! error TS2322:             Type '(...args: [] | [PromiseLike<undefined>]) => Promise<IteratorResult<string, void>>' is not assignable to type '(...args: [] | [PromiseLike<undefined>]) => Promise<IteratorResult<number, any>>'.
+!!! error TS2322:             Type '(...args: [] | [undefined]) => Promise<IteratorResult<string, void>>' is not assignable to type '(...args: [] | [undefined]) => Promise<IteratorResult<number, any>>'.
 !!! error TS2322:               Type 'Promise<IteratorResult<string, void>>' is not assignable to type 'Promise<IteratorResult<number, any>>'.
         yield "a";
     };
@@ -128,7 +128,7 @@ tests/cases/conformance/types/asyncGenerators/types.asyncGenerators.es2018.2.ts(
 !!! error TS2322:       Type '() => AsyncGenerator<string, void, unknown>' is not assignable to type '() => AsyncIterator<number, any, undefined>'.
 !!! error TS2322:         Type 'AsyncGenerator<string, void, unknown>' is not assignable to type 'AsyncIterator<number, any, undefined>'.
 !!! error TS2322:           Types of property 'next' are incompatible.
-!!! error TS2322:             Type '(...args: [] | [unknown]) => Promise<IteratorResult<string, void>>' is not assignable to type '(...args: [] | [PromiseLike<undefined>]) => Promise<IteratorResult<number, any>>'.
+!!! error TS2322:             Type '(...args: [] | [unknown]) => Promise<IteratorResult<string, void>>' is not assignable to type '(...args: [] | [undefined]) => Promise<IteratorResult<number, any>>'.
 !!! error TS2322:               Type 'Promise<IteratorResult<string, void>>' is not assignable to type 'Promise<IteratorResult<number, any>>'.
         yield* (async function * () { yield "a"; })();
     };
@@ -211,7 +211,7 @@ tests/cases/conformance/types/asyncGenerators/types.asyncGenerators.es2018.2.ts(
                                              ~~~~~~~~~~~~~~~~
 !!! error TS2322: Type 'AsyncGenerator<number, any, undefined>' is not assignable to type 'Iterator<number, any, undefined>'.
 !!! error TS2322:   Types of property 'next' are incompatible.
-!!! error TS2322:     Type '(...args: [] | [PromiseLike<undefined>]) => Promise<IteratorResult<number, any>>' is not assignable to type '(...args: [] | [undefined]) => IteratorResult<number, any>'.
+!!! error TS2322:     Type '(...args: [] | [undefined]) => Promise<IteratorResult<number, any>>' is not assignable to type '(...args: [] | [undefined]) => IteratorResult<number, any>'.
 !!! error TS2322:       Type 'Promise<IteratorResult<number, any>>' is not assignable to type 'IteratorResult<number, any>'.
 !!! error TS2322:         Property 'value' is missing in type 'Promise<IteratorResult<number, any>>' but required in type 'IteratorYieldResult<number>'.
 !!! related TS2728 /.ts/lib.es2015.iterable.d.ts:33:5: 'value' is declared here.

--- a/tests/cases/compiler/customAsyncIterator.ts
+++ b/tests/cases/compiler/customAsyncIterator.ts
@@ -1,0 +1,13 @@
+// @target: esnext
+
+// GH: https://github.com/microsoft/TypeScript/issues/33239
+class ConstantIterator<T> implements AsyncIterator<T, undefined, T | undefined> {
+    constructor(private constant: T) {
+    }
+    async next(value?: T): Promise<IteratorResult<T>> {
+        if (value != null) {
+            throw new Error("ConstantIterator.prototype.next may not take any values");
+        }
+        return { value: this.constant, done: false };
+    }
+}

--- a/tests/cases/conformance/types/asyncGenerators/types.asyncGenerators.es2018.1.ts
+++ b/tests/cases/conformance/types/asyncGenerators/types.asyncGenerators.es2018.1.ts
@@ -124,3 +124,6 @@ async function * awaitedType1() {
 async function * awaitedType2() {
     const x = await Promise.resolve(1);
 }
+async function * nextType1(): { next(...args: [] | [number | PromiseLike<number>]): any } {
+    const x = yield; // `number | PromiseLike<number>` (should not await TNext)
+}

